src/Protocol.cpp

```diff
@@ -0,0 +1,277 @@
+    m_configuration.phase_lock_enabled = data[46];
+    m_configuration.phase_lock_angle = parsePhaseLockAngle(data + 47);
+    m_configuration.time_utc = parseUTC(data);
+
+    std::vector<int> prism_angles;
+    for (int i = 0; i < 5; i++) {
+        short int prism_angle = data[240 + i * 2] * 256 + data[241 + i * 2];
+        prism_angles.emplace_back(std::move(prism_angle));
+    }
+    getOffsetAngle(prism_angles);
+}
+
+base::samples::Pointcloud Protocol::getPointCloud()
+{
+    return m_point_cloud;
+}
+
+base::Point Protocol::getPoint(uint8_t line_number,
+    double horizontal_angle_degree,
+    double distance)
+{
+    double vertical_angle_sin =
+        computeVerticalAngleSin(line_number, horizontal_angle_degree);
+    double vertical_angle_cos = sqrt(1 - vertical_angle_sin * vertical_angle_sin);
+    base::Point point;
+    point.x() = distance * vertical_angle_cos * cos(toRad(horizontal_angle_degree));
+    point.y() = distance * vertical_angle_cos * sin(toRad(horizontal_angle_degree));
+    point.z() = distance * vertical_angle_sin;
+    return point;
+}
+
+base::Vector4d colorByReflectivity(uint8_t intensity)
+{
+    if (intensity < 30) {
+        return base::Vector4d(0,
+            static_cast<int>(intensity * 255 / 30) & 0xff,
+            0xff,
+            0xff);
+    }
+    else if (intensity < 90) {
+        return base::Vector4d(0,
+            0xff,
+            static_cast<int>((90 - intensity) * 255 / 60) & 0xff,
+            0xff);
+    }
+    else if (intensity < 150) {
+        return base::Vector4d(static_cast<int>((intensity - 90) * 255 / 60) & 0xff,
+            0xff,
+            0,
+            0xff);
+    }
+    else {
+        return base::Vector4d(0xff,
+            static_cast<int>((255 - intensity) * 255 / (256 - 150)) & 0xff,
+            0,
+            0xff);
+    }
+}
+
+std::optional<base::samples::Pointcloud> Protocol::handleSingleEcho(unsigned char* data)
+{
+    if (m_full_frame == true) {
+        m_point_cloud.points.clear();
+        m_point_cloud.colors.clear();
+        m_full_frame = false;
+    }
+    uint64_t timestamp_microseconds =
+        (static_cast<uint64_t>(data[1200] << 24) +
+            static_cast<uint64_t>(data[1201] << 16) +
+            static_cast<uint64_t>(data[1202] << 8) + data[1203]);
+    int miliseconds = timestamp_microseconds / 1000;
+    int microseconds = timestamp_microseconds % 1000;
+    base::Time time = base::Time::fromTimeValues(m_configuration.time_utc.year,
+        m_configuration.time_utc.month,
+        m_configuration.time_utc.day,
+        data[1197],
+        data[1198],
+        data[1199],
+        miliseconds,
+        microseconds);
+    for (int i = 0; i < 1197; i = i + 7) {
+        if (data[i] == 0xff && data[i + 1] == 0xaa && data[i + 2] == 0xbb &&
+            data[i + 3] == 0xcc && data[i + 4] == 0xdd) {
+            m_full_frame = true;
+            m_point_cloud.time = time;
+            return m_point_cloud;
```

Comment:
In the packets containing the start marker, there is no points (all the distances are 0), so there is no problem to ignore the last bytes here